lib/Models/ApiTableCatalogItem.ts

```diff
@@ -0,0 +1,197 @@
+
+const automaticTableStylesStratumName = TableAutomaticStylesStratum.stratumName;
+/**
+ * THE API AND TRAITS OF THIS EXPERIMENTAL CATALOG ITEM SHOULD BE CONSIDERED IN
+ * ALPHA. EXPECT BREAKING CHANGES.
+ *
+ * This is a generic, one-size-fits-most catalog item for deriving tables from
+ * external APIs. Currently only supports JSON APIs, and doesn't support region
+ * mapping. Also currently only supports a single API to get values from, and a
+ * single API to get positions from.
+ */
+export class ApiTableCatalogItem extends AutoRefreshingMixin(
+  TableMixin(CatalogMemberMixin(CreateModel(ApiTableCatalogItemTraits)))
+) {
+  static readonly type = "api-table";
+  @observable protected apiResponses: any[] = [];
+  @observable protected hasData: boolean = false;
+
+  constructor(id: string | undefined, terria: Terria) {
+    super(id, terria);
+    this.strata.set(
+      automaticTableStylesStratumName,
+      new TableAutomaticStylesStratum(this)
+    );
+  }
+
+  @computed
+  get apiDataIsLoaded(): boolean {
+    return this.apiResponses.length > 0;
+  }
+
+  @computed
+  get apiDataColumnMajor() {
+    return this.apiResponseToTable();
+  }
+
+  protected loadDataFromApis() {
+    return Promise.all(
+      this.apis.map(async api => {
+        const data = await loadJson(proxyCatalogItemUrl(this, api.apiUrl!));
+        return Promise.resolve({
+          data,
+          api
+        });
+      })
+    ).then((values: { data: any[]; api: Model<ApiTraits> }[]) => {
+      runInAction(() => {
+        // Make map of ids to values that are constant for that id
+        const perIdData: Map<string, any> = new Map(
+          values
+            .filter(val => val.api.kind === "PER_ID") // per id only
+            .map(val => val.data) // throw away api, keep data
+            .reduce((curr, prev) => curr.concat(prev), []) // flatten
+            // make id/data pair for perIdData map
+            .map(data => [data[this.idKey!], data])
+        );
+
+        // Merge all responses with the same id (and timestamp, if applicable),
+        // Each of these merged responses will become a row in the table.
+        const perRowData = values
+          .filter(val => val.api.kind === "PER_ROW")
+          .map(val => val.data)
+          .reduce((curr, prev) => curr.concat(prev), []);
+
+        // Group all responses wth the same id
+        const groupedData = groupBy(perRowData, this.idKey);
+
+        this.apiResponses = Object.keys(groupedData).map(key => {
+          // Then, merge all responses with that key into a single object
+          const mergedResponse = groupedData[key].reduce(
+            (prev, curr) => Object.assign(prev, curr),
+            {}
+          );
+
+          // Add per id data for this response
+          Object.assign(
+            mergedResponse,
+            perIdData.get(mergedResponse[this.idKey!])
+          );
+          return mergedResponse;
+        });
+      });
+    });
```

Comment:
Ideally there would be no observables in this before setting `apiResponses` - but we can't do this until `TableMixin` has been sorted